File: x-baseproject/src/main/java/com/xia/baseproject/rxhttp/callback/Callback.java
package com.xia.baseproject.rxhttp.callback;

import android.arch.lifecycle.LifecycleOwner;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.Disposable;
import okhttp3.ResponseBody;

/**
 * @author xia
 * @date 2018/7/27.
 */
@SuppressWarnings("WeakerAccess")
public abstract class Callback<T> {
    private WeakReference<Context> mContext;
    private WeakReference<LifecycleOwner> mLifecycleOwner;

    public Context getContext() {
        return mContext.get();
    }

    public LifecycleOwner getLifecycleOwner() {
        return mLifecycleOwner.get();
    }

    public Callback(@NonNull LifecycleOwner lifecycleOwner) {
        mLifecycleOwner = new WeakReference<>(lifecycleOwner);
        if (lifecycleOwner instanceof AppCompatActivity) {
            mContext = new WeakReference<>((AppCompatActivity) lifecycleOwner);
        } else if (lifecycleOwner instanceof Fragment) {
            mContext = new WeakReference<>(((Fragment) lifecycleOwner).getContext());
        }
    }

    public void onSubscribe(Disposable d) {
    }

    public void onError(String error) {
    }

    public void onComplete() {
    }

    public void inProgress(float progress, long total) {
    }

    public abstract T parseNetworkResponse(ResponseBody responseBody) throws Exception;

    public abstract void onResponse(T response);
}

File: fly-http/src/main/java/com/xia/fly/http/callback/Callback.kt
package com.xia.fly.http.callback

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.LifecycleOwner
import io.reactivex.disposables.Disposable
import okhttp3.ResponseBody
import java.lang.ref.WeakReference

/**
 * @author xia
 * @date 2018/7/27.
 */
@Suppress("UNUSED_PARAMETER")
abstract class Callback<T>(private val lifecycleOwner: LifecycleOwner) {
    private var mContext: WeakReference<Context>? = null

    init {
        var context: Context? = null
        if (lifecycleOwner is AppCompatActivity) {
            context = lifecycleOwner
        } else if (lifecycleOwner is Fragment) {
            context = lifecycleOwner.context
        }
        context?.let {
            mContext = WeakReference(it)
        }
    }

    fun getContext(): Context? {
        return mContext?.get()
    }

    fun getLifecycleOwner(): LifecycleOwner {
        return lifecycleOwner
    }

    fun onSubscribe(d: Disposable) {}

    fun onError(e: Throwable) {}

    fun onComplete() {}

    fun inProgress(progress: Float, total: Long) {}

    @Throws(Exception::class)
    abstract fun parseNetworkResponse(responseBody: ResponseBody)

    abstract fun onResponse(response: T?)
}
